### PR TITLE
Hood sim

### PR DESCRIPTION
**Pull request by rakrakon**

### CODE REVIEW

File: src/main/java/frc/robot/subsystems/hood/Hood.java [hunk 1]
```diff
@@ -49,6 +51,10 @@ public MutableMeasure<Angle> getAngle() {
         return inputs.angle;
     }
 
+    public MutableMeasure<Voltage> getPower() {
```
Comment (vichik123) [inline src/main/java/frc/robot/subsystems/hood/Hood.java]: This function should be called `getVoltage`.

File: src/test/java/frc/robot/subsystems/example/HoodTest.java [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+    public HoodTest() {
+        HoodIO hoodIO = new HoodIOSim();
+        Hood.initialize(hoodIO);
+        hood = Hood.getInstance();
+    }
+
+    @Test
+    public void testSetAngle() throws InterruptedException {
+        hood.setDefaultCommand(hood.setAngle(() -> Units.Radians.of(0.5).mutableCopy()));
+        Thread.sleep(3000);
```
Comment (vichik123) [inline src/test/java/frc/robot/subsystems/example/HoodTest.java]: Same as the shooter, you need to call the `CommandScheduler.getInstance.run()` before sleeping.

File: src/test/java/frc/robot/subsystems/example/HoodTest.java [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+        Assertions.assertTrue(hood.atSetpoint());
+    }
+
+    @Test
+    public void testSetPower() throws InterruptedException {
+        hood.setDefaultCommand(hood.setPower(() -> 0.5));
+        Thread.sleep(3000);
```
Comment (vichik123) [inline src/test/java/frc/robot/subsystems/example/HoodTest.java]: Same here.

File: src/test/java/frc/robot/subsystems/example/HoodTest.java [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+        Assertions.assertEquals(0.5, hood.getPower().in(Units.Volts), Delta);
```
Comment (vichik123) [inline src/test/java/frc/robot/subsystems/example/HoodTest.java]: If you set a power of 0.5, the motor will run at 50% of the maximal voltage, which is 6.0 volts. I'm not sure it will run exactly at 6.0 volts however, so I don't know how valid this test is.

**Review by vichik123 — CHANGES_REQUESTED** [review]

File: src/test/java/frc/robot/subsystems/example/HoodTest.java (hunk 4 above)
Comment (vichik123) [reply src/test/java/frc/robot/subsystems/example/HoodTest.java]: Well actually, I know that it will work, which is why I think this test isn't especially effective.

File: src/test/java/frc/robot/subsystems/example/HoodTest.java (hunk 2 above)
Comment (vichik123) [reply src/test/java/frc/robot/subsystems/example/HoodTest.java]: I just realized that the update of the simulated motor needs to be called.  I think this means that the test has to run on a different thread.

Comment (rakrakon) [reply src/test/java/frc/robot/subsystems/example/HoodTest.java]: there is an example of how to run the test it in a different thread in what katzuv sent i think
https://github.com/wpilibsuite/allwpilib/blob/main/wpilibjExamples/src/test/java/edu/wpi/first/wpilibj/examples/elevatorsimulation/ElevatorSimulationTest.java

File: src/main/java/frc/robot/subsystems/hood/Hood.java (hunk 1 above)
Comment (rakrakon) [reply src/main/java/frc/robot/subsystems/hood/Hood.java]: 7b8b479

File: src/test/java/frc/robot/subsystems/example/HoodTest.java (hunk 3 above)
Comment (rakrakon) [reply src/test/java/frc/robot/subsystems/example/HoodTest.java]: eaf9038

File: src/test/java/frc/robot/subsystems/example/HoodTest.java (hunk 4 above)
Comment (rakrakon) [reply src/test/java/frc/robot/subsystems/example/HoodTest.java]: 75ee8e8
How is it not effective it tests if the setpower function reaches its setpoint

File: src/test/java/frc/robot/subsystems/example/HoodTest.java (hunk 2 above)
Comment (rakrakon) [reply src/test/java/frc/robot/subsystems/example/HoodTest.java]: 61c8b29
i used SimHooks.stepTiming
which advances the simulation time forward

**Review by katzuv — CHANGES_REQUESTED** [review]
Sorry, I forgot about that – missing logging the robot-relative position of the shooter. @vichik123 can help if needed.

File: src/test/java/frc/robot/subsystems/example/HoodTest.java (hunk 4 above)
Comment (vichik123) [reply src/test/java/frc/robot/subsystems/example/HoodTest.java]: Because behind the scenes, the applied voltage is the calculation for it. That means that if the motor uses position control, the applied voltage would be the PID calculation for it. And if it's simply voltage it would be that voltage. It's like that also for the real motor I believe.

**Comment by vichik123** [comment]
> Sorry, I forgot about that – missing logging the robot-relative position of the shooter. @vichik123 can help if needed.

@rakrakon Here is a file depicting the places of the cameras, and the poses of the robot parts are on the bottom.
https://docs.google.com/document/d/1KjueG1yS05T2dkQ523dc6AQl4ikLDm4vGbyrLCavjAc/edit

File: src/test/java/frc/robot/subsystems/example/HoodTest.java (hunk 4 above)
Comment (rakrakon) [reply src/test/java/frc/robot/subsystems/example/HoodTest.java]: ok, so remove the test?

Comment (vichik123) [reply src/test/java/frc/robot/subsystems/example/HoodTest.java]: Yeah I think it's best.

Comment (katzuv) [reply src/test/java/frc/robot/subsystems/example/HoodTest.java]: It can also be modified to test the angle setting and not the power.

Comment (rakrakon) [reply src/test/java/frc/robot/subsystems/example/HoodTest.java]: There is already a different test for that

Comment (rakrakon) [reply src/test/java/frc/robot/subsystems/example/HoodTest.java]: eb1c199

**Comment by rakrakon** [comment]
> > Sorry, I forgot about that – missing logging the robot-relative position of the shooter. @vichik123 can help if needed.
> 
> @rakrakon Here is a file depicting the places of the cameras, and the poses of the robot parts are on the bottom. https://docs.google.com/document/d/1KjueG1yS05T2dkQ523dc6AQl4ikLDm4vGbyrLCavjAc/edit

There is no hood Pose there

File: src/main/java/frc/robot/subsystems/hood/Hood.java [hunk 5]
```diff
@@ -70,12 +79,23 @@ public Command updateInternalEncoder() {
...
     /** Updates the state of the hood. */
     @Override
     public void periodic() {
         io.updateInputs();
         Logger.processInputs("Hood", inputs);
 
         hood.setAngle(inputs.angle.in(Units.Degrees));
+
+        Logger.recordOutput("HoodPose", getPose3d(inputs.angle.in(Units.Degrees)));
+        SmartDashboard.putData("HoodMech", mechanism2d);
```
Comment (katzuv) [inline src/main/java/frc/robot/subsystems/hood/Hood.java]: The mechanism should instead be `@AutoLogOutput`.

File: src/main/java/frc/robot/subsystems/hood/Hood.java [hunk 6]
```diff
@@ -70,12 +79,23 @@ public Command updateInternalEncoder() {
...
+
     /** Updates the state of the hood. */
     @Override
     public void periodic() {
         io.updateInputs();
         Logger.processInputs("Hood", inputs);
 
         hood.setAngle(inputs.angle.in(Units.Degrees));
+
+        Logger.recordOutput("HoodPose", getPose3d(inputs.angle.in(Units.Degrees)));
```
Comment (katzuv) [inline src/main/java/frc/robot/subsystems/hood/Hood.java]: You can add `@AutoLogOutput` to functions, too.

File: src/main/java/frc/robot/subsystems/hood/Hood.java [hunk 7]
```diff
@@ -49,6 +54,10 @@ public MutableMeasure<Angle> getAngle() {
         return inputs.angle;
     }
 
+    public MutableMeasure<Voltage> getVoltage() {
+        return inputs.voltage;
+    }
+
```
Comment (katzuv) [inline src/main/java/frc/robot/subsystems/hood/Hood.java]: Why is that needed?

File: src/main/java/frc/robot/subsystems/hood/HoodConstants.java [hunk 8]
```diff
@@ -1,12 +1,46 @@
...
+import frc.robot.Constants;
+import lib.webconstants.LoggedTunableNumber;
 
 public class HoodConstants {
     public static Translation2d MECHANISM_2D_POSE = new Translation2d(1, 1);
-    public static double HOOD_LENGTH = 0.4; // m
+    public static Measure<Distance> HOOD_LENGTH = Units.Meters.of(0.4);
     public static Measure<Dimensionless> POSITION_TOLERANCE = Units.Percent.of(5);
+    public static double MAX_VELOCITY = 1; // RPS
+    public static double MAX_ACCELERATION = 4; // RPS squared
```
Comment (katzuv) [inline src/main/java/frc/robot/subsystems/hood/HoodConstants.java]: Measure

File: src/main/java/frc/robot/subsystems/hood/HoodConstants.java [hunk 9]
```diff
@@ -1,12 +1,46 @@
...
+    public static final double GEAR_RATIO = 1.0;
+    public static final double MOMENT_OF_INERTIA = 0.0003;
+    public static final double HOOD_X = -0.27;
+    public static final double HOOD_Y = 0.2385;
+    public static final Translation2d ROOT_POSITION = new Translation2d(HOOD_X, HOOD_Y);
```
Comment (katzuv) [inline src/main/java/frc/robot/subsystems/hood/HoodConstants.java]: Unless `HOOD_X` and `HOOD_Y` are used somewhere else than `ROOT_POSITION`, you can just write the values directly in `ROOT_POSITION`'s decleration.

File: src/main/java/frc/robot/subsystems/hood/HoodConstants.java [hunk 10]
```diff
@@ -1,12 +1,46 @@
...
 
 public class HoodConstants {
     public static Translation2d MECHANISM_2D_POSE = new Translation2d(1, 1);
-    public static double HOOD_LENGTH = 0.4; // m
+    public static Measure<Distance> HOOD_LENGTH = Units.Meters.of(0.4);
     public static Measure<Dimensionless> POSITION_TOLERANCE = Units.Percent.of(5);
+    public static double MAX_VELOCITY = 1; // RPS
+    public static double MAX_ACCELERATION = 4; // RPS squared
+    public static final double GEAR_RATIO = 1.0;
+    public static final double MOMENT_OF_INERTIA = 0.0003;
```
Comment (katzuv) [inline src/main/java/frc/robot/subsystems/hood/HoodConstants.java]: Measure or at least add units in a comment.

**Review by katzuv — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/subsystems/hood/HoodConstants.java (hunk 8 above)
Comment (vichik123) [reply src/main/java/frc/robot/subsystems/hood/HoodConstants.java]: It can't be a measure, there is no rotational acceleration unit.

Comment (katzuv) [reply src/main/java/frc/robot/subsystems/hood/HoodConstants.java]: `Measure<Velocity<Velocity<Angle>>` doesn't work?

Comment (katzuv) [reply src/main/java/frc/robot/subsystems/hood/HoodConstants.java]: Oh you need a unit. You can use the `Per` method tho

File: src/main/java/frc/robot/subsystems/hood/Hood.java (hunk 6 above)
Comment (rakrakon) [reply src/main/java/frc/robot/subsystems/hood/Hood.java]: cb100d9
25f3b05

File: src/main/java/frc/robot/subsystems/hood/HoodConstants.java (hunk 9 above)
Comment (rakrakon) [reply src/main/java/frc/robot/subsystems/hood/HoodConstants.java]: a2e31af

File: src/main/java/frc/robot/subsystems/hood/HoodConstants.java (hunk 10 above)
Comment (rakrakon) [reply src/main/java/frc/robot/subsystems/hood/HoodConstants.java]: d980235

File: src/main/java/frc/robot/subsystems/hood/HoodConstants.java (hunk 8 above)
Comment (rakrakon) [reply src/main/java/frc/robot/subsystems/hood/HoodConstants.java]: 4d4a202

File: src/main/java/frc/robot/subsystems/hood/Hood.java (hunk 7 above)
Comment (rakrakon) [reply src/main/java/frc/robot/subsystems/hood/Hood.java]: It was needed for the test we removed
0d14d35

File: src/main/java/frc/robot/subsystems/hood/Hood.java (hunk 5 above)
Comment (rakrakon) [reply src/main/java/frc/robot/subsystems/hood/Hood.java]: that was already the case, just removed the smartDahsboard.putData
a2f3916

File: src/main/java/frc/robot/subsystems/hood/Hood.java (hunk 6 above)
Comment (katzuv) [reply src/main/java/frc/robot/subsystems/hood/Hood.java]: 1. "update" is not a correct term, "get" really is more accurate. This function returns a value, it doesn't update anything.
2. I don't think you need to call the function by yourself, AdvantageKit does it by itself.

File: src/main/java/frc/robot/subsystems/hood/HoodConstants.java (hunk 10 above)
Comment (katzuv) [reply src/main/java/frc/robot/subsystems/hood/HoodConstants.java]: Awesome!

File: src/main/java/frc/robot/subsystems/hood/Hood.java (hunk 6 above)
Comment (rakrakon) [reply src/main/java/frc/robot/subsystems/hood/Hood.java]: 57219ac

**Review by vichik123 — APPROVED** [review]
